Allow numerical fluxes to retrieve volume_tags

nilsvu opened the pull request:
## Proposed changes

_Edited on Sep 4, 2019:_ This PR now creates an `interface` interface from the metafunctions that were in `Domain/Tags.hpp` and uses it to treat numerical fluxes similar to interface compute items.

_Edited on May 31, 2019:_ The linear elasticity system operators need access to the `ConstitutiveRelation` (describing the material properties), which is going to be available through the DataBox once PDALs #1535 are merged. However, all `argument_tags` for the flux operators are currently taken from the element interface. So this PR adds a `volume_tags` typelist that works the same as it does for interface compute items, i.e. it specifies the `argument_tags` that should be taken directly from the DataBox instead of the interface.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu commented:
@wthrowe @nilsdeppe do you think this approach makes sense, or can you think of a better way to make available the `ConstitutiveRelation` (which is created from options) to the operators?

wthrowe commented:
I didn't look at the details of the second commit, but this looks like a good way to go to me.

nilsvu commented:
Ok @wthrowe @nilsdeppe I propagated this to all systems and tests, this is ready for review

wthrowe commented on src/Elliptic/Initialization/BoundaryConditions.hpp:
```diff
@@ -67,43 +68,57 @@ namespace Initialization {
   using sources_tag =
       db::add_tag_prefix<Tags::Source, typename system::fields_tag>;
+  using numerical_flux_tag = typename Metavariables::normal_dot_numerical_flux;
+  using numerical_flux_type = typename numerical_flux_tag::type;
```
`NumericalFluxType` or `NumericalFlux` (The tag is fine as is if I remember our conventions correctly.)

wthrowe commented on src/Elliptic/Systems/Poisson/Equations.hpp:
```diff
@@ -187,6 +190,9 @@ struct FirstOrderInternalPenaltyFlux {
       noexcept;
 
+  using boundary_argument_tags =
+      tmpl::list<Tags::Normalized<Tags::UnnormalizedFaceNormal<Dim>>>;
```
This addition seems unrelated to the rest of the commit.

nilsdeppe reviewed (dismissed):
If this is only necessary because you con't get cache tags through the DataBox then I'm quite strongly against it. This isn't a small hack, this is a major hack that's probably more work than actually getting the right thing to work...

nilsvu commented:
The tags taken from the DataBox are sliced to the interface. But here I need to retrieve an item from the ConstGlobalCache.

nilsdeppe commented:
Yes, but you can grab volume tags for interface things too so that no slicing is done, as far as I'm aware.

nilsvu commented:
I tried different ways of working with interface tags and DataBox and settled on this design, seems cleanest to me. Happy to discuss on Monday.

nilsdeppe commented:
Let's do that because if all that's necessary is getting cache tags from the DataBox then I'd prefer that route :)

nilsvu commented:
Conclusion from today's discussion was to see if we can get phase-dependent action lists in soon, which will enable retrieval of cache tags through the DataBox. Then I can refactor this PR, specifying a `volume_tags` list for the flux quantities.

nilsvu commented:
@nilsdeppe @wthrowe I made the changes to this PR that we discussed a long time ago. This PR doesn't actually depend on PDALs, but just adds the `volume_tags` functionality to the flux operators.

wthrowe commented:
Could this be done generically by adding `mutate_apply`, etc., wrappers that act on interfaces?  I think the helpers for `InterfaceComputeItem` can already do most of what is needed.

nilsvu commented:
@wthrowe I refactored the changes to make better use of existing infrastructure:
- The elliptic N.F are now just interface compute items, I'll add them in another PR.
- For the numerical fluxes I use helper code from interface compute items. Once we have PDALs, we can possibly refactor that to also just be a compute item.
- The elliptic boundary conditions code is different enough so that the compute item code is not useful. That will need to be refactored once we have general boundary conditions, but until then I think he current code works fine.

wthrowe reviewed (changes requested):
You can't use the private `Interface_detail` functions outside of the interface implementation.  Is there a reason a generic `interface_mutate_apply` function or similar added to the (SpECTRE) interface (C++) interface would not work?

nilsvu committed:
Rename InterfaceComputeItem to InterfaceCompute

nilsvu committed:
Move interface compute tags into separate file

nilsvu committed:
Clean up InterfaceHelpers interface

nilsvu committed:
Allow numerical fluxes to retrieve volume_tags

nilsvu commented:
@nilsdeppe @wthrowe I just updated this (quite old) PR to accomplish what it was intended for, i.e. allow numerical fluxes to specify `volume_tags` similar to interface compute items, by following @wthrowe's suggestion above. This PR now does the following:
- Moves the interface compute tags (not the simple tags, so only affects code that _adds_ them to the DataBox) out of `Domain/Tags.hpp` (which had grown to 649 lines already) into `Domain/InterfaceComputeTags.hpp` and the interface helper metafunctions to `Domain/InterfaceHelpers.hpp`
- Renames `Tags::InterfaceComputeItem<...>` to `Tags::InterfaceCompute<>` following our recent conventions
- Uses the (now public) `InterfaceHelpers.hpp` to treat numerical fluxes like interface compute items. I didn't find that adding an `interface_mutate_apply` function (as @wthrowe suggested) would help, since numerical fluxes have a quite particular interface with their `package_data` function and changing that interface is out of scope for this PR I think.

Could you review and let me know your thoughts on this please?

wthrowe commented:
This isn't compiling.  Also, GitHub appears to be showing me the commits in the wrong order.  What order should they be in?

nilsvu committed:
Fix various includes

nilsvu commented:
@wthrowe It wasn't compiling because of missing includes in unrelated parts of the code that are showing up now because I'm removing a bunch of unneeded includes from `Domain/Tags.hpp`. Travis will probably turn up more and I'll squash them into the "Fix various includes" commit.

This is the commit order (from `develop` to `HEAD`):

- Fix various includes
- Move interface compute tags into separate file
- Rename InterfaceComputeItem to InterfaceCompute
- Clean up InterfaceHelpers interface
- Allow numerical fluxes to retrieve volume_tags

nilsvu commented:
@nilsdeppe @wthrowe could you take a look at this PR please?

wthrowe commented:
Don't make the interface helpers public.  Do something like
```c++
interface_apply<DirectionsTag, tmpl::list<Tags::Mesh, ArgumentTags...>, VolumeTags>(
  [&normal_dot_numerical_flux_computer](Bla&mesh, Bla&... args) noexcept {
    PackagedData ret(...);
    n_dot_f_computer.package_data(make_not_null(&ret), args...);
    return ret;
  },
  box);
```
(I haven't  really thought about what the correct template arguments, etc., are.)  That should return a `FixedHashMap<Direction, PackagedData>`.  This does all the directions at once, as opposed to doing them separately as is currently done, but all uses of this function just call it in a loop over directions ayway.

nilsvu committed:
Add interface_apply function

This function applies an invokable to the DataBox on interfaces.
It allows to retrieve data from the DataBox on a set of interfaces.

nilsvu committed:
Add volume_dim alias to direction tags

nilsvu commented:
@wthrowe thanks for the suggestion, the approach works quite well. Please take a look at my implementation :)

In addition to the two fixups, the following commits are new:
- Add interface_apply function
- Add volume_dim alias to direction tags

nilsvu commented:
@wthrowe please take a look when you get the chance

wthrowe commented on src/Domain/InterfaceHelpers.hpp:
```diff
@@ -0,0 +1,119 @@
+
+/// Add `::Tags::Interface` wrappers to the tags if they are not listed as being
+/// taken from the volume.
+template <typename Tag, typename DirectionsTag>
```
Please leave the descriptive type name here.  There are a lot of tags flying around in this code, so anything that requires a particular one of them should indicate that.  I think all the other changed functions are clear just from the name, but `make_interface_argument_tags` doesn't have an obvious argument meaning.

nilsvu replied:
you're right, the name wasn't particularly good. This function actually doesn't care what kind of type it works with, just retrieves its `argument_tags` and `volume_tags`. So named this the more general `get_interface_argument_tags<InterfaceInvokable, DirectionsTag>` now. Could also make this public (haven't yet). Thoughts?

wthrowe replied:
That name is better.

I can't think of a use for this outside of interface compute items (and maybe an interface_apply overload).  What do you want a public version for?

nilsvu replied:
nothing in particular comes to mind, let's leave it detail for now then

wthrowe commented on src/Domain/InterfaceHelpers.hpp:
```diff
@@ -0,0 +1,119 @@
+                                   ::Tags::Interface<DirectionsTag, Tag>>;
+};
+
+/// Add `::Tags::Interface` wrappers to the tags if they are not listed as being
```
"the tags" -> "the argument tags" or similar.  Also no doxygen.

wthrowe commented on src/Domain/InterfaceHelpers.hpp:
```diff
@@ -0,0 +1,119 @@
+    const db::DataBox<DbTagsList>& box, tmpl::list<ArgumentTags...> /*meta*/,
+    ExtraArgs&&... extra_args) noexcept {
+  using interface_return_type = std::decay_t<decltype(
+      interface_invokable(std::declval<db::item_type<ArgumentTags>>()...,
```
I think this can be made to work with `db::BaseTag`s (at least for the volume tags).  See if adding the second argument to `db::item_type` does that.  If not don't worry about it.

nilsvu replied:
yep, works for volume tags

wthrowe commented on tests/Unit/Domain/Test_InterfaceHelpers.cpp:
```diff
@@ -0,0 +1,51 @@
+    const ElementId<1> element_id{0, {{{1, 0}}}};
+    const ElementId<1> east_id{0, {{{1, 1}}}};
+    const Element<1> element{element_id,
+                             {{Direction<1>::upper_xi(), {{east_id}, {}}}}};
```
Do a test with more than one direction and one with no directions.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/InterfaceActionHelpers.hpp:
```diff
@@ -10,89 +10,80 @@
+      DirectionsTag,
+      tmpl::flatten<tmpl::list<::Tags::Mesh<VolumeDim - 1>,
+                               typename NumericalFlux::argument_tags>>,
+      InterfaceHelpers_detail::get_volume_tags<NumericalFlux>>(
```
Move `get_volume_tags` out of the detail namespace so you're not accessing private stuff here.  I think that function is clear/useful enough for general use.

nilsvu commented:
@wthrowe pushed two more fixups "fixup InterfaceHelpers" and "fixup interface_apply and tests". Please take a look.

wthrowe commented on tests/Unit/Domain/Test_InterfaceHelpers.cpp:
```diff
@@ -16,36 +16,81 @@ struct SomeNumber : db::SimpleTag {
+  test_interface_apply<1, ::Tags::InternalDirections<1>>(
+      // Reference element no internal directions:
+      // [ | X | | ]-> xi
+      {{0, {{{2, 1}}}}, {}}, {}, {});
```
Shouldn't no internal directions be `[ X ]` and `{0, 0}`?

nilsvu replied:
yes, fixed.

wthrowe reviewed (commented):
Squash.

nilsvu commented:
@sxs-collaboration/spectre-core-devs please review or assign another reviewer
